### PR TITLE
Add sms channel

### PR DESCRIPTION
**Pull request by dlozina**
- [x] Added SMS channel

- [x] Added Integration tests for SMS channel

- [x] Added model tests for SMS channel

### CODE REVIEW

File: infobip_channels/sms/models/body/core.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    SOCIAL_INVITES = "SOCIAL_INVITES"
+
+
+class TimeUnitEnum(str, Enum):
+    MINUTE = "MINUTE"
+    HOUR = "HOUR"
+    DAY = "DAY"
+
+
+class DaysEnum(str, Enum):
```
Comment (kilicluka) [inline infobip_channels/sms/models/body/core.py:42]: We have this exact enum in `mms` channel also, might be good to move it into a `core` models file and reuse it.

File: infobip_channels/sms/models/body/core.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    to: constr(min_length=0, max_length=50)
+
+
+class DeliveryTimeWindow(CamelCaseModel):
+    days: conlist(DaysEnum, min_items=1)
+    from_time: Optional[Time] = Field(alias="from", default=None)
+    to: Optional[Time] = None
+
+    @root_validator
+    def validate_from_and_to(cls, values):
```
Comment (kilicluka) [inline infobip_channels/sms/models/body/core.py:77]: I believe this is the same method as we have in the `mms` channel. We can extract it (along with `_validate_time_differences`) into a `FromAndToTimeValidator` and reuse it in both places.

File: infobip_channels/sms/models/body/reschedule_sms_messages.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+from pydantic import validator
+
+from infobip_channels.core.models import MessageBodyBase
+
+
+class RescheduleSMSMessagesMessageBody(MessageBodyBase):
+    send_at: Optional[Union[datetime, str]] = None
+
+    @validator("send_at")
+    def convert_send_at_to_correct_format(cls, value):
```
Comment (kilicluka) [inline infobip_channels/sms/models/body/reschedule_sms_messages.py:13]: This is very similar to the `convert_send_at_to_correct_format` method we use in the `mms` channel. We could extract the common logic into a Validator class and change this version to:

```python
class RescheduleSMSMessagesMessageBody(Validator, MessageBodyBase):
    ...
    @validator("send_at")
    def convert_send_at_to_correct_format(cls, value):
            if value > datetime.now() + timedelta(days=180):
                raise ValueError(
                    "Scheduled message must me sooner than 180 days from today"
                )
            return super().convert_send_at_to_correct_format(value)
```

File: infobip_channels/sms/channel.py [hunk 4]
```diff
@@ -0,0 +1,360 @@
...
+            message.dict(by_alias=True),
+            PostHeaders(authorization=self._client.auth.api_key),
+        )
+        return self._construct_response(response, PreviewSMSMessageResponse)
+
+    def get_outbound_sms_delivery_reports(
+        self,
+        query_parameters: Union[
+            GetOutboundSMSDeliveryReportsQueryParameters, Dict
+        ] = {},
```
[the remark this answers is not in the document]
Comment (dlozina) [reply infobip_channels/sms/channel.py:203]: top top!

File: infobip_channels/core/models.py [hunk 5]
```diff
@@ -48,6 +65,37 @@ def validate_url_length(cls, value: str) -> str:
...
+        return date_time_format.strftime("%Y-%m-%dT%H:%M:%S.%fZ")
+
+    @classmethod
+    def convert_time_to_correct_format_validate_limit(cls, value):
+
+        date_time_format = cls.convert_to_date_time_format(value)
+
+        if date_time_format > datetime.now() + timedelta(days=cls._MAX_TIME_LIMIT):
+            raise ValueError(
+                "Scheduled message must me sooner than 180 days from today"
```
[the remark this answers is not in the document]
Comment (dlozina) [reply infobip_channels/core/models.py:93]: 👍🏻

File: infobip_channels/core/models.py [hunk 6]
```diff
@@ -48,6 +65,37 @@ def validate_url_length(cls, value: str) -> str:
...
+            )
+        else:
```
[the remark this answers is not in the document]
Comment (dlozina) [reply infobip_channels/core/models.py:95]: 👍🏻

File: infobip_channels/core/models.py [hunk 7]
```diff
@@ -48,6 +65,37 @@ def validate_url_length(cls, value: str) -> str:
...
+        if isinstance(value, str):
+            value = datetime.fromisoformat(value)
+
+        return value
+
+    @classmethod
+    def convert_time_to_correct_format(cls, value) -> str:
+        date_time_format = cls.convert_to_date_time_format(value)
+
+        return date_time_format.strftime("%Y-%m-%dT%H:%M:%S.%fZ")
```
[the remark this answers is not in the document]
Comment (dlozina) [reply infobip_channels/core/models.py:84]: 👍🏻

File: infobip_channels/sms/channel.py [hunk 8]
```diff
@@ -0,0 +1,360 @@
...
+
+
+class SMSChannel(Channel):
+    """Class used for interaction with the Infobip's SMS API."""
+
+    SMS_URL_TEMPLATE_VERSION_1 = "/sms/1/"
+    SMS_URL_TEMPLATE_VERSION_2 = "/sms/2/"
+
+    @staticmethod
+    def validate_query_parameter(
```
[the remark this answers is not in the document]
Comment (dlozina) [reply infobip_channels/sms/channel.py:81]: 👍🏻

File: infobip_channels/sms/channel.py [hunk 9]
```diff
@@ -0,0 +1,360 @@
...
+
+    def _get_custom_response_class(
+        self,
+        raw_response: Union[requests.Response, Any],
+        response_class: Type[ResponseBase] = SendSMSResponse,
+        *args,
+        **kwargs
+    ) -> Type[ResponseBase]:
+
+        if raw_response.status_code in (HTTPStatus.OK,):
```
[the remark this answers is not in the document]
Comment (dlozina) [reply infobip_channels/sms/channel.py:106]: oh, this was bad 😞

File: infobip_channels/sms/channel.py [hunk 10]
```diff
@@ -0,0 +1,360 @@
...
+
+        :param message: Body of the message to send
+        :return: Received response
+        """
+        message = self.validate_message_body(message, SMSMessageBody)
+
+        response = self._client.post(
+            self.SMS_URL_TEMPLATE_VERSION_2 + "text/advanced",
+            message.dict(by_alias=True),
+            PostHeaders(authorization=self._client.auth.api_key),
```
[the remark this answers is not in the document]
Comment (dlozina) [reply infobip_channels/sms/channel.py:137]: Yea, used MMS as a boiler plate. Thanks

File: infobip_channels/sms/channel.py [hunk 11]
```diff
@@ -0,0 +1,360 @@
...
+
+        response = self._client.get(
+            self.SMS_URL_TEMPLATE_VERSION_1 + "reports",
+            params=query_parameters.dict(by_alias=True),
+        )
+        return self._construct_response(response, OutboundDeliveryReportsResponse)
+
+    def get_outbound_sms_message_logs(
+        self,
+        query_parameters: Union[GetOutboundSMSLogsQueryParameters, Dict] = {},
```
[the remark this answers is not in the document]
Comment (dlozina) [reply infobip_channels/sms/channel.py:226]: 🗡️

File: infobip_channels/sms/channel.py [hunk 12]
```diff
@@ -0,0 +1,360 @@
...
+
+        response = self._client.get(
+            self.SMS_URL_TEMPLATE_VERSION_1 + "logs",
+            params=query_parameters.dict(by_alias=True),
+        )
+        return self._construct_response(response, OutboundMessageLogsResponse)
+
+    def get_inbound_sms_messages(
+        self,
+        query_parameters: Union[GetInboundSMSMessagesQueryParameters, Dict] = {},
```
[the remark this answers is not in the document]
Comment (dlozina) [reply infobip_channels/sms/channel.py:248]: ⚔️

File: infobip_channels/sms/channel.py [hunk 13]
```diff
@@ -0,0 +1,360 @@
...
+        )
+
+        message = self.validate_message_body(
+            message, UpdateScheduledSMSMessagesMessageBody
+        )
+
+        response = self._client.put(
+            self.SMS_URL_TEMPLATE_VERSION_1 + "bulks/status",
+            message.dict(by_alias=True),
+            PutHeaders(authorization=self._client.auth.api_key),
```
[the remark this answers is not in the document]
Comment (dlozina) [reply infobip_channels/sms/channel.py:354]: 👍🏻

File: infobip_channels/sms/models/body/core.py (hunk 1 above)
Comment (dlozina) [reply infobip_channels/sms/models/body/core.py:42]: done

File: infobip_channels/sms/models/body/core.py (hunk 2 above)
Comment (dlozina) [reply infobip_channels/sms/models/body/core.py:77]: yea, good chunk of code is gone

File: infobip_channels/sms/models/body/reschedule_sms_messages.py (hunk 3 above)
Comment (dlozina) [reply infobip_channels/sms/models/body/reschedule_sms_messages.py:13]: done done, thanks

**Pull request merged**